### PR TITLE
889 - Add automated testing specific to TYDE

### LINKED ISSUE #889: Add automated testing specific to TYDE

### PR DESCRIPTION
**Pull request by rlnsy**
## Changes
1. Adds automated tests related to the following:
- Modules
- Stages
- Subtopics / Question sets
- Completion / progress for each of the above
-  Cockpit functions (favorite, profile, clicking on modules)
2. Also makes some source code tweaks to allow for testing, including a small extension to the tapestry API.
## Limitations
- Had troubles with jest tests, so the tests here are purely end-to-end. I'm open to suggestions for elements that would be better tested in Vue testing.
- Code coverage is quite good, but there are some random bits of code that I could not get to. Certain components could use more testing in the future, namely the activities summary (I was blocked on this due to a creation bug)
## Issue Linkage
Closes #889 
## PR Dependency
N/A
## Testing/Reviewing
Would love suggestions on test cases, code quality, optimizations etc.

### CODE REVIEW

File: templates/vue/cypress/integration/tyde.spec.js
Comment (shirleywbi) [inline templates/vue/cypress/integration/tyde.spec.js:75]: Submitted a bug report for this but I think there's an issue with getSelectedNode since it seems to be selecting Root, instead of the question set.
image

**Review by shirleywbi — COMMENTED** [review]
@rlnsy I think you have the major behaviour captured. The only thing I can think of is whether we need a test for the profile page.

File: endpoints.php
Comment (rlnsy) [reply endpoints.php:523]: Yes it is. There was a bug where virtually all of my fixtures were being rejected, and including this extra field is a quick hack to make it work.

File: templates/vue/cypress/integration/tyde.spec.js
Comment (rlnsy) [reply templates/vue/cypress/integration/tyde.spec.js:75]: Oh good catch! Super easy to get false positives with cypress. I can actually just convert this to get the node by title.

File: templates/vue/cypress/support/commands.js
Comment (rlnsy) [reply templates/vue/cypress/support/commands.js:206]: I think something like open menu might work.

File: templates/vue/src/components/node-modal/SpaceshipPartForm.vue
Comment (rlnsy) [reply templates/vue/src/components/node-modal/SpaceshipPartForm.vue:6]: Hmm it should be but it seems to be working with this old version somehow. I'll make the change.

Comment (rlnsy) [reply templates/vue/src/components/node-modal/SpaceshipPartForm.vue:6]: @shirleywbi I changed it and it broke - I think because `input-test-id` is used specifically for text entry fields, so I'm gonna use this for now.

File: templates/vue/cypress/integration/tyde.spec.js
Comment (rlnsy) [inline templates/vue/cypress/integration/tyde.spec.js:292]: @wynnset This assertion fails right now. Can you confirm what is actually supposed to happen here? After that is figured out, this is ready for review.

Comment (wynnset) [reply templates/vue/cypress/integration/tyde.spec.js:292]: @rlnsy In the profile, the answer the user provided to the given question is supposed to show up. If you are answering using audio, then an audio element should exist with the recorded audio. Here you are checking for the question, but I think you need to be checking for the answer.

**Review by shirleywbi — APPROVED** [review]
LGTM!

**Comment by rlnsy** [comment]
@shirleywbi @wynnset It looks like the failing test above (after merging the base branch) is an actual break in functionality that my tests detected. That is, I am unable to add a subtopic node to a question set when testing locally as well. Should I file this as a bug on Ansana?

**Comment by shirleywbi** [comment]
@rlnsy If it's a functionality issue and not a test issue, please file the bug. Thank you.

**Comment by wynnset** [comment]
@rlnsy I agree with @shirleywbi . Could you please update the status of this task?

File: templates/vue/cypress/integration/tyde.spec.js
Comment (wynnset) [inline templates/vue/cypress/integration/tyde.spec.js:25]: Could we use some images that are from a more reliable source like Wikimedia?

**Review by wynnset — CHANGES_REQUESTED** [review]
Mostly looking good. Just one minor change... Please mark as `ready to merge` when changes are committed.

File: templates/vue/cypress/integration/tyde.spec.js
Comment (rlnsy) [reply templates/vue/cypress/integration/tyde.spec.js:25]: Good call

**Review by wynnset — APPROVED** [review]

**Pull request merged**
